src/Tools/dotnet-gcdump/DotNetHeapDump/EventPipeDotNetHeapDumper.cs

```diff
@@ -0,0 +1,238 @@
+                    {
+                        break;
+                    }
+
+                    if (!etwDataPresent && sw.Elapsed.TotalSeconds > 5)      // Assume it started within 5 seconds.  
+                    {
+                        log.WriteLine("{0,5:n1}s: Assume no Dot Heap", sw.Elapsed.TotalSeconds);
+                        break;
+                    }
+                    if (sw.Elapsed.TotalSeconds > 20)       // Time out after 100 seconds. 
+                    {
+                        log.WriteLine("{0,5:n1}s: Timed out after 20 seconds", sw.Elapsed.TotalSeconds);
+                        break;
+                    }
+                    // TODO FIX NOW, time out faster if we seek to be stuck
+                    if (dumpComplete)
+                    {
+                        break;
+                    }
+                }
+
+                gcDumpSession.EndSession();
+
+                while (!readerTask.Wait(1000))
+                    Console.WriteLine("still reading...");
+                if (etwDataPresent)
+                {
+                    dumper.ConvertHeapDataToGraph();        // Finish the conversion.  
+                }
+            }
+            finally
+            {
+                // Stop the ETW providers
+                log.WriteLine("{0,5:n1}s: Shutting down EventPipe session", sw.Elapsed.TotalSeconds);
+            }
+            if (readerTask != null)
+            {
+                log.WriteLine("{0,5:n1}s: Waiting for shutdown to complete.", sw.Elapsed.TotalSeconds);
+                if (!readerTask.Wait(2000))
+                {
+                    log.WriteLine("{0,5:n1}s: Shutdown wait timed out after 2 seconds.", sw.Elapsed.TotalSeconds);
+                }
+            }
+            log.WriteLine("[{0,5:n1}s: Done Dumping .NET heap success={1}]", sw.Elapsed.TotalSeconds, dumpComplete);
+
+            return dumpComplete;
+        }
+    }
+
+    internal class EventPipeSession
+    {
+        private List<Provider> _providers;
+        private Stream _eventPipeStream;
+        private EventPipeEventSource _source;
+        private ulong _sessionId;
+        private int _pid;
+
+        public ulong SessionId => _sessionId;
+        public IReadOnlyList<Provider> Providers => _providers.AsReadOnly();
+        public EventPipeEventSource Source => _source;
+
+        public EventPipeSession(int pid, List<Provider> providers, bool requestRundown = true)
+        {
+            _pid = pid;
+            _providers = providers;
+
+            if (!requestRundown)
+            {
+                var config = new SessionConfigurationV2(
+                    circularBufferSizeMB: 1024,
+                    format: EventPipeSerializationFormat.NetTrace,
+                    requestRundown: false,
+                    providers
+                );
+                _eventPipeStream = EventPipeClient.CollectTracing2(pid, config, out _sessionId);
```

Comment:
You could use CollectTracing2 all the time, supporting a 3.0 Preview 6 runtime is unlikely to matter.